src/Sarif/EnumeratedArtifact.cs

```diff
@@ -16,44 +16,162 @@ public EnumeratedArtifact(IFileSystem fileSystem)
+                }
+                return this.encoding;
+            }
+
+            set => this.encoding = value;
+        }
 
         internal IFileSystem FileSystem { get; set; }
 
+        /// <summary>
+        /// Gets or sets a property that determines whether the enumerated artifact
+        /// can obtain data from a non-seekable stream. This operation will be
+        /// less performant in any case where an artifact is a textual file (as
+        /// the file will first be converted to a byte array to determine
+        /// whether it is textual and subsequently turned into a string if it is).
+        /// </summary>
+        public bool SupportNonSeekableStreams { get; set; }
+
         public string Contents
         {
-            get => GetContents();
+            get => GetArtifactData().text;
             set => this.contents = value;
         }
 
-        private string GetContents()
+        public byte[] Bytes
+        {
+            get => GetArtifactData().bytes;
+            set => this.bytes = value;
+        }
+
+        private (string text, byte[] bytes) GetArtifactData()
         {
-            if (this.contents != null) { return this.contents; }
+            if (this.contents != null)
+            {
+                return (this.contents, bytes: null);
+            }
+
+            if (this.bytes != null)
+            {
+                return (text: null, this.bytes);
+            }
 
-            if (Stream == null && this.contents == null)
+            if (Stream == null && this.contents == null && this.bytes == null)
             {
-                // TBD we actually have no validation URI is non-null yet.
-                contents = Uri!.IsFile
-                    ? FileSystem.FileReadAllText(Uri.LocalPath)
-                    : null;
+                if (Uri == null || (Uri.IsAbsoluteUri && !Uri.IsFile))
+                {
+                    throw new InvalidOperationException();
+                }
 
-                this.sizeInBytes = (long?)this.contents?.Length;
+                // This is our client-side, disk-based file retrieval case.
+                Stream = FileSystem.FileOpenRead(Uri.LocalPath);
+            }
+
+            if (Stream.CanSeek)
+            {
+                RetrieveDataFromSeekableStream();
             }
             else
             {
-                if (Stream.CanSeek) { this.Stream.Seek(0, SeekOrigin.Begin); }
+                RetrieveDataFromNonSeekableStream();
+            }
+
+            Stream = null;
+
+            return (this.contents, this.bytes);
+        }
+
+        private void RetrieveDataFromNonSeekableStream()
+        {
+            bool isText;
+            if (!SupportNonSeekableStreams)
+            {
+                throw new InvalidOperationException("Stream is not seekable. Provide a seekable stream or set the 'SupportNonSeekableStreams' property.");
+            }
+
+            this.bytes = new byte[Stream.Length];
```

Comment:
Why is it the case that we can categorize a Nonseekable stream in 4096 bytes, but we need to read in the full stream for a Seekable one?  If 4096 is sound, I suspect we'd both love to take a one-block peek in the Nonseekable case, and use that as the starting buffer for a full Stream or byte[] read depending on what is appropriate.  So I probably just don't see a fatal flaw yet.

Where is 4096 coming from in the Seekable case?  Looking at the implementation of FileEncoding, I'd optimistically credit that we probably don't get much more accuracy in the classifier after this many bytes if that's coming from some tribal knowledge or guidance stemming from other users of this code.  But right now that's just a brittle prior without any experience on with so many of these file extensions, so I'm really curious what the author guidance or community discourse looks like.

Reply:
Here's maybe a way to put a finger on what I'm getting at:

The snippet code we're taking asserts that if you see no more than 10% of the bytes as being non-textual, then that's a good-enough heuristic.  I buy that.  Not second-guessing it.  But that's the kind of suspiciously-round number someone just yanked out of their brain one day and it just worked well enough in the real world that it stuck and became an institution.

Maybe 4096 is the same way and we don't need to do full file scans to classify in any scenario?  This is what I'm wondering/hoping.  (Not asserting.)

Reply:
OK, so first, 4096 is *way* too much data and I dropped it to 1024. This is probably *still* more data than we need.

Let's talk offline further about this (in particular about whether we can address the inefficiency in the non-seekable case of reading the whole thing). I'm just not aware of how we can handle a string decoding from the original non-seekable string if we've already obtained 1024 bytes from it.

Reply:
Actually: if the zip archive is always already in memory, maybe closing/reopening is trivially cheap?